File: clojure/src/aoc/point.clj
(ns aoc.point
  "2D point operations and direction handling for Advent of Code.")

;; Direction constants
(def NORTH [0 -1])
(def SOUTH [0 1])
(def EAST [1 0])
(def WEST [-1 0])
(def NORTH-EAST [1 -1])
(def NORTH-WEST [-1 -1])
(def SOUTH-EAST [1 1])
(def SOUTH-WEST [-1 1])

(def CARDINAL-DIRECTIONS
  "Four cardinal directions: N, E, S, W"
  [NORTH EAST SOUTH WEST])

(def ALL-DIRECTIONS
  "All eight directions including diagonals"
  [NORTH NORTH-EAST EAST SOUTH-EAST SOUTH SOUTH-WEST WEST NORTH-WEST])

(defn point
  "Create a point from x and y coordinates.

   Parameters:
   - x: X coordinate
   - y: Y coordinate

   Returns: Vector [x y]"
  [x y]
  [x y])

(defn add
  "Add two points together.

   Parameters:
   - p1: First point [x y]
   - p2: Second point [x y]

   Returns: New point [x y]"
  [[x1 y1] [x2 y2]]
  [(+ x1 x2) (+ y1 y2)])

(defn subtract
  "Subtract second point from first.

   Parameters:
   - p1: First point [x y]
   - p2: Second point [x y]

   Returns: New point [x y]"
  [[x1 y1] [x2 y2]]
  [(- x1 x2) (- y1 y2)])

(defn scale
  "Multiply a point by a scalar.

   Parameters:
   - p: Point [x y]
   - n: Scalar value

   Returns: New point [x y]"
  [[x y] n]
  [(* x n) (* y n)])

(defn manhattan-distance
  "Calculate Manhattan distance between two points.

   Parameters:
   - p1: First point [x y]
   - p2: Second point [x y]

   Returns: Integer distance"
  [[x1 y1] [x2 y2]]
  (+ (Math/abs (- x1 x2))
     (Math/abs (- y1 y2))))

(defn move
  "Move a point in a given direction.

   Parameters:
   - point: Starting point [x y]
   - direction: Direction vector [dx dy]
   - steps: Number of steps (default 1)

   Returns: New point [x y]"
  ([point direction]
   (move point direction 1))
  ([point direction steps]
   (add point (scale direction steps))))

(defn neighbors-4
  "Get the 4 cardinal neighbors of a point.

   Parameters:
   - p: Point [x y]

   Returns: Sequence of neighboring points"
  [p]
  (map #(add p %) CARDINAL-DIRECTIONS))

(defn neighbors-8
  "Get all 8 neighbors of a point (including diagonals).

   Parameters:
   - p: Point [x y]

   Returns: Sequence of neighboring points"
  [p]
  (map #(add p %) ALL-DIRECTIONS))

(defn rotate-left
  "Rotate a direction vector 90 degrees counter-clockwise.

   Parameters:
   - direction: Direction vector [dx dy]

   Returns: Rotated direction vector"
  [[dx dy]]
  [dy (- dx)])

(defn rotate-right
  "Rotate a direction vector 90 degrees clockwise.

   Parameters:
   - direction: Direction vector [dx dy]

   Returns: Rotated direction vector"
  [[dx dy]]
  [(- dy) dx])

(defn reverse-direction
  "Reverse a direction vector.

   Parameters:
   - direction: Direction vector [dx dy]

   Returns: Reversed direction vector"
  [[dx dy]]
  [(- dx) (- dy)])

(defn in-bounds?
  "Check if a point is within grid boundaries.

   Parameters:
   - point: Point [x y]
   - width: Grid width
   - height: Grid height

   Returns: Boolean"
  [[x y] width height]
  (and (>= x 0) (< x width)
       (>= y 0) (< y height)))

File: go/pkg/point/point.go
// Package point provides 2D point types and operations for grid-based puzzles.
package point

import "fmt"

// Point represents a 2D coordinate (x, y) or (col, row).
type Point struct {
	X, Y int
}

// New creates a new Point with the given coordinates.
func New(x, y int) Point {
	return Point{X: x, Y: y}
}

// Add returns the sum of two points (vector addition).
func (p Point) Add(other Point) Point {
	return Point{X: p.X + other.X, Y: p.Y + other.Y}
}

// Sub returns the difference of two points (vector subtraction).
func (p Point) Sub(other Point) Point {
	return Point{X: p.X - other.X, Y: p.Y - other.Y}
}

// Mul returns the point scaled by a scalar value.
func (p Point) Mul(scalar int) Point {
	return Point{X: p.X * scalar, Y: p.Y * scalar}
}

// Manhattan returns the Manhattan distance between two points.
func (p Point) Manhattan(other Point) int {
	dx := p.X - other.X
	dy := p.Y - other.Y
	if dx < 0 {
		dx = -dx
	}
	if dy < 0 {
		dy = -dy
	}
	return dx + dy
}

// ManhattanFromOrigin returns the Manhattan distance from the origin.
func (p Point) ManhattanFromOrigin() int {
	return p.Manhattan(Point{0, 0})
}

// String returns a string representation of the point.
func (p Point) String() string {
	return fmt.Sprintf("(%d,%d)", p.X, p.Y)
}

// Neighbors4 returns the 4 orthogonal neighbors (up, down, left, right).
func (p Point) Neighbors4() []Point {
	return []Point{
		p.Add(North),
		p.Add(South),
		p.Add(East),
		p.Add(West),
	}
}

// Neighbors8 returns all 8 neighbors (including diagonals).
func (p Point) Neighbors8() []Point {
	return []Point{
		p.Add(North),
		p.Add(NorthEast),
		p.Add(East),
		p.Add(SouthEast),
		p.Add(South),
		p.Add(SouthWest),
		p.Add(West),
		p.Add(NorthWest),
	}
}

// RotateLeft rotates the point 90 degrees counterclockwise around the origin.
func (p Point) RotateLeft() Point {
	return Point{X: p.Y, Y: -p.X}
}

// RotateRight rotates the point 90 degrees clockwise around the origin.
func (p Point) RotateRight() Point {
	return Point{X: -p.Y, Y: p.X}
}

// Direction vectors for the four cardinal directions.
// In grid coordinates: North is up (negative Y), South is down (positive Y).
var (
	North = Point{X: 0, Y: -1}
	South = Point{X: 0, Y: 1}
	East  = Point{X: 1, Y: 0}
	West  = Point{X: -1, Y: 0}
)

// Direction vectors for the eight cardinal and diagonal directions.
var (
	NorthEast = Point{X: 1, Y: -1}
	SouthEast = Point{X: 1, Y: 1}
	SouthWest = Point{X: -1, Y: 1}
	NorthWest = Point{X: -1, Y: -1}
)

// Cardinal4 returns all 4 cardinal directions.
var Cardinal4 = []Point{North, East, South, West}

// Cardinal8 returns all 8 directions (cardinal + diagonal).
var Cardinal8 = []Point{North, NorthEast, East, SouthEast, South, SouthWest, West, NorthWest}

// TurnLeft returns the direction after turning left (counterclockwise).
func TurnLeft(dir Point) Point {
	switch dir {
	case North:
		return West
	case West:
		return South
	case South:
		return East
	case East:
		return North
	default:
		return dir.RotateLeft()
	}
}

// TurnRight returns the direction after turning right (clockwise).
func TurnRight(dir Point) Point {
	switch dir {
	case North:
		return East
	case East:
		return South
	case South:
		return West
	case West:
		return North
	default:
		return dir.RotateRight()
	}
}

// Reverse returns the opposite direction.
func Reverse(dir Point) Point {
	return Point{X: -dir.X, Y: -dir.Y}
}

// ParseDirection converts a character to a direction vector.
// Accepts: 'N', 'S', 'E', 'W', 'U', 'D', 'L', 'R', '^', 'v', '<', '>'
func ParseDirection(ch rune) Point {
	switch ch {
	case 'N', 'U', '^':
		return North
	case 'S', 'D', 'v':
		return South
	case 'E', 'R', '>':
		return East
	case 'W', 'L', '<':
		return West
	default:
		panic(fmt.Sprintf("unknown direction: %c", ch))
	}
}
